### PR TITLE
Better views

### PR DESCRIPTION
**Pull request by XertroV**
- use materialized views
- fix n+1 issue with requesting children

* use ARHQ to generate 'recursive' queries on-the-fly
* *MUCH* fasteer than using node_decendants_incr (10-1000x)
* FIXED: Generating pages for nodes close to root (e.g. /0, /19; 600ms) takes longer than for deeper nodes (e.g. /100; 50ms) -- using 88k faked nodes
  * bottleneck seems to be related to joining to nodes_readable; not sure if there's a way to speed that up
  * esp as we want to apply permissions early in the SQL query
  * mb we could add a term in the recursive bit?
    * <https://github.com/GeorgeKaraszi/ActiveRecordExtended#overlap>
    * yup!
* disabled node_descendants_incr because it wasn't useful, but it didn't seem to speed up inserts (88k nodes still takes ~3 min)
  * maybe we should enable it again in case it becomes useful? (since it's built up incrementally) -- low overhead
* the main outcome is `Node.children_rec_arhq` and things are built around that.
  * arhq -> activerecord-heirarchical-query (the lib used to provide .join_recursive)
  * rec -> "recursive"
  * it was named this because it's easy to grep for, plus it let me work and compare to other impls I had, like `.children_rec_arel` which was written with more arel stuff.

## structure of views: (roughtly; there are some queries done in models/node.rb for efficiency)

Legend: 
* **F**: filter
* **J**: join
* orange: column names
* *with_desc* -> "with descendants" -> this is basically the `Node.children_rec_arhq` method

### CODE REVIEW

**Comment by XertroV** [comment]
commits today are just tidying

**Comment by XertroV** [comment]
updated the PR description a bit too

File: app/controllers/authors_controller.rb [hunk 1]
```diff
@@ -11,5 +13,7 @@ def mine
 
   def set_author
     @author = Author.find(params[:id])
+    @posts = Node.get_nodes_readable_by(current_user) { |q| where("author_id = #{author.id}") }
```
Comment (dev28237582) [inline app/controllers/authors_controller.rb]: Breaks all calls to the Authors controller. The `get_nodes_readable_by` method has been removed from `Node`.

Comment (dev28237582) [reply app/controllers/authors_controller.rb]: Fix submitted https://github.com/fi-tools/cf-forum/pull/19/commits/b3572b6f0295cc18d29dc6d9e91254623ba6fc9b

**Review by decentgarlic — APPROVED** [review]
lgtm 👍 (re the most recent commit)

**Comment by dev28237582** [comment]
merging into `dev` as this is essentially `dev` now. :laughing:

**Pull request merged**